k9mail/src/main/java/com/fsck/k9/controller/imap/ImapSync.java

```diff
@@ -399,61 +391,62 @@ public void messageRemoved(Message message) {
         return newMessages.get();
     }
 
-    private void evaluateMessageForDownload(final Message message, final String folder,
-            final LocalFolder localFolder,
-            final Folder remoteFolder,
-            final List<Message> unsyncedMessages,
-            final List<Message> syncFlagMessages,
-            boolean flagSyncOnly,
-            SyncListener listener) throws MessagingException {
+    private void evaluateMessageForDownload(Message message, BackendFolder backendFolder, Folder remoteFolder,
+            List<Message> unsyncedMessages, List<Message> syncFlagMessages, boolean flagSyncOnly) {
+
+        String messageServerId = message.getUid();
         if (message.isSet(Flag.DELETED)) {
-            Timber.v("Message with uid %s is marked as deleted", message.getUid());
+            Timber.v("Message with uid %s is marked as deleted", messageServerId);
 
             syncFlagMessages.add(message);
             return;
         }
 
-        LocalMessage localMessage = localFolder.getMessage(message.getUid());
-
-        if (localMessage == null) {
+        boolean messagePresentLocally = backendFolder.isMessagePresent(messageServerId);
+        if (!messagePresentLocally) {
             if (!flagSyncOnly) {
-                if (!message.isSet(Flag.X_DOWNLOADED_FULL) && !message.isSet(Flag.X_DOWNLOADED_PARTIAL)) {
-                    Timber.v("Message with uid %s has not yet been downloaded", message.getUid());
-
-                    unsyncedMessages.add(message);
-                } else {
-                    Timber.v("Message with uid %s is partially or fully downloaded", message.getUid());
```

Comment:
where'd this part go? not needed anymore because of the stuff below?

Reply:
`ImapStore` never returns messages with either `Flag.X_DOWNLOADED_FULL` or `Flag.X_DOWNLOADED_PARTIAL` set. I removed the branch so I didn't have to add additional methods to `BackendFolder`.